drivers/fpga/dfl.c

```diff
@@ -1213,7 +1213,7 @@ static int dfh_get_param_size(void __iomem *dfh_base, resource_size_t max)
 	return -ENOENT;
 }
 
-#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 0, 0)
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 0, 0) && RHEL_RELEASE_CODE < 0x808
```

Comment:
I downloaded the source for the 4.18.0-477.10.1.el8_8 kernel to look at the code more carefully. I can see that the unrolled_memcpy_fromio() function is present, but the fix for the performance issue is not present in that kernel (e.g. arch/x86/lib/iomem.c does not exist). I believe that if we take this change as is, we will re-introduce the performance issue for RHEL 8.8 and above (until RedHat picks up the performance fix).

The original #ifdef is bounding two things:
(1) The definition of unrolled_memcpy_fromio()
(2) The use of unrolled_memcpy_fromio() as a performance workaround.

I think for RHEL 8.8 we want to use the workaround, but we don't want to redefine the function. We can probably do that making the "RHEL_RELEASE_CODE < 0x808" case a nested #if that bounds only the function definition.

Reply:
We do want to use our own function also for RHEL 8.8. The use of the name unrolled_memcpy_fromio() was just inspired by the newer  kernels, its not a precise copy of the original anyway (using readq() the orig used readb()). We can call it what we want, e.g. backport_memcpy_fromio(). Then we can avoid doing something special for RHEL 8.8 and above.

Reply:
Also note that performance issue is an understatement. The driver is hanging for 40s because the reading fails because the card does not allow byte reading. So it does not work without using the special function.

Reply:
Ugh, that's huge! I guess then renaming and using the local version would be the only feasible approach.

Is there a way I can replicate the test?

Reply:
If you have an Arrow Creek card, N6010 or N6011 and use RHEL (or Rocky Linux) 8.x. Just load the drivers.
You might also need firmware image to be a beta 2 release. I am using image build by Silicom and I shared it with Russ, but I do not know if he did see the error, because Ubuntu 22.04 he used, use a different kernel version.

Reply:
I did see the error. I found a lab machine with RHEL8 on it and was able to reproduce the error there.

Reply:
I agree that simply renaming the function to avoid the name conflict is the right approach. I had done a visual comparison of the conflicting functions but didn't notice the readq vs readb differences.